src/psbt/mod.rs

```diff
@@ -861,6 +800,160 @@ impl PsbtExt for Psbt {
     }
 }
 
+/// Extension trait for PSBT inputs
+pub trait PsbtInputExt {
+    /// Given the decriptor for utxo being spent, populate the input with all the required fields to
+    /// spend it.
+    ///
+    /// If the descriptor contains wilcards or otherwise cannot be transformed into a concrete
+    /// descriptor an error will returned.
+    ///
+    /// You can make sure the descriptor is consistent with the `witness_utxo` and/or
+    /// `non_witness_utxo` fields of the PSBT by setting `check_witness_utxo` and
+    /// `check_non_witness_utxo` respectively. You should set `check_non_witness_utxo` to inputs the
```

Comment:
> Is there any special reason why you want unchecked APIs? I don't think there is significant efficiency gain here.
>
> Normally, you can use the prevout information in the current transaction to find this `vout`. I think we should mandate the user to have this field instead of an `Option<u32>` one and always do the check.

The optionality of the checks here are not about efficiency but rather about being flexible with respect to the interpretation of BIP174.

Do you want to make it mandatory to have `non_witness_utxo` for segwitv0 outputs? i.e. Do we want to stop people being able to voluntarily expose themselves to the segwit bug? In the existing implementation of `update_desc` this check is optional. It is simply only turned on if `non_witness_utxo` is present and `witness_utxo` is omitted (note that this is not quite right for other reasons -- you could set `witness_utxo` for a non-segwit output and it would pass the check here). To keep this "feature" it has to be an optional check. FWIW I would be fine with dropping this misfeature. BDK currently exposes this feature too (but has the check on by default) but I think we could remove it. (cc @afilini)

I want the previous index `u32` to be a param rather than extracting it from the PSBT because I want a function that goes from `UTXO-data -> psbt::Input` without having to construct a PSBT to do that. Does that make sense? As you mention in the future this will work perfectly with `PsbtV2`. So I propose to keep the functionality on `psbt::Input` (which requires the `u32` for now) and then just drop the argument when moving to v2 (or change it to a bool depending on how the discussion relating to the previous paragraph turns out). 

So finally, what about at least making the `witness_utxo` check mandatory? I think this makes sense as long as there are no applications that simply omit `witness_utxo` and only include `non_witness_utxo`. [BIP174] does not say that `witness_utxo` is mandatory if you have already included `non_witness_utxo`. It says you can have both but doesn't say that. Do we want to be opinionated here and say you always have to include `witness_utxo` if it's segwit?

Having considered the above I think there is a bug here where you can omit `non_witness_utxo` if it's taproot without providing `witness_utxo`. I'll fix that now.

[BIP174]: https://github.com/bitcoin/bips/blob/master/bip-0174.mediawiki

Reply:
Good point about the segwit change calculation bug. I don't think that having this field in psbt helps prevent this bug? We also need to check the txid calculation for guard against it. Also, this is only really useful for HW wallets use-case. 

Code sample from BIP 174. 
```
    if non_witness_utxo.exists:
        assert(sha256d(non_witness_utxo) == psbt.tx.input[i].prevout.hash)
```

There are perhaps more gotchas that users will have to be aware of while signing transactions of inputs that they don't have all inputs. I think there are also other attacks where HW wallets need to be careful while signing only a partial set of inputs of a psbt that we cannot prevent. 

Note that only HW wallets might need to have this check as they cannot confirm the amounts against the blockchain. Wallets that can watch blockchain don't need this field. Secondly, this field is super big and we should avoid it if it is not necessary. How about we make two versions `update_with_descriptor` and `update_with_descriptor_unchecked`? In the former, we can mandate `non_witness_utxo` is correct. 

> I want the previous index u32 to be a param rather than extracting it from the PSBT because I want a function that goes from UTXO-data -> psbt::Input without having to construct a PSBT to do that. Does that make sense?

Indeed. I think we are on the same page.

Reply:
Yes you're right we should check the prevout matches the txid otherwise we are not guaranteeing anything much. It still helps to make sure it's there so the hardware wallet can check it but perhaps this code will make it onto a hardware wallet someday :)

I think the `checked` and `unchecked` idea is good. I'll try making  the method on the PSBT itself the fully checked version since it can check the TXID against the prevout and the one directly on the `psbt::Input` the unchecked version? This can be revisited once we have psbtv2.